hash_map.h

```diff
@@ -0,0 +1,164 @@
+//
+// Created by semyon on 03.02.2020
+//
+
+#ifndef LONG_CONTEST_HASHMAP_H
+#define LONG_CONTEST_HASHMAP_H
+
+#include <memory>
+#include <vector>
+#include <list>
+
+template<class KeyType, class ValueType, class Hash = std::hash<KeyType> >
+class HashMap {
+ public:
+  using iterator = typename std::list<std::pair<const KeyType, ValueType>>::iterator;
+  using const_iterator = typename std::list<std::pair<const KeyType, ValueType>>::const_iterator;
+  using pair_type = std::pair<const KeyType, ValueType>;
+  using pair_type_nonconst = std::pair<KeyType, ValueType>;
+  const std::size_t size_default = 1024;
+
+  iterator begin() noexcept {
+      return nodes.begin();
+  }
+
+  const_iterator begin() const noexcept {
+      return nodes.begin();
+  }
+
+  iterator end() noexcept {
+      return nodes.end();
+  }
+
+  const_iterator end() const noexcept {
+      return nodes.end();
+  }
+
+  HashMap(Hash hasher = Hash()) : size_table(size_default), hasher(hasher), table(size_default) {}
+
+  template<class IteratorType>
+  HashMap(IteratorType begin_, IteratorType end_, Hash hasher = Hash()) : HashMap() {
+      for (auto &i = begin_; i != end_; i++) {
+          insert(*i);
+      }
+  }
+
+  HashMap(std::initializer_list<pair_type_nonconst> init, Hash hasher = Hash()) : HashMap(init.begin(), init.end()) {}
+
+  std::size_t size() const {
+      return nodes.size();
+  }
+
+  bool empty() const {
+      return nodes.empty();
+  }
+
+  const Hash hash_function() const {
+      return hasher;
+  }
+
+  iterator insert(pair_type insertable) {
+      const KeyType user_key = insertable.first;
+      if (find(user_key) != nodes.end()) {
+          return find(user_key);
+      }
+      const std::size_t cur_hash = hasher(user_key) % size_table;
+      nodes.push_back(insertable);
+      const iterator cur_it = std::prev(nodes.end(), 1);
+      table[cur_hash].push_back(cur_it);
+      if (table[cur_hash].size() > size_table / size_default) {
+          rebuilt();
+      }
+      return cur_it;
+
+  }
+
+  void erase(KeyType user_key) {
+      const std::size_t cur_hash = hasher(user_key) % size_table;
+      if (find(user_key) != nodes.end()) {
```

Comment:
Does this find necessary?

Looks like you can just iterate over all bucket, if `find(user_key) == nodes.end()`.